File: stat_react/server/publicNiftySpot.mjs
/**
 * Fallback NIFTY 50 when mStock Type B is unavailable (e.g. expired JWT).
 * Spot + 1m candles from the same Yahoo ^NSEI response so chart and headline match.
 */

const CHART_URL =
  'https://query1.finance.yahoo.com/v8/finance/chart/%5ENSEI?interval=1m&range=1d';
const DAILY_CHART_URL =
  'https://query1.finance.yahoo.com/v8/finance/chart/%5ENSEI?interval=1d&range=5d';

function finite(v) {
  if (typeof v === 'number' && Number.isFinite(v)) return v;
  return null;
}

/**
 * @param {unknown} r
 * @returns {{ ltp: number | null, bars: { time: number, open: number, high: number, low: number, close: number }[], error: string, note?: string }}
 */
function parseYahooChartResult(r) {
  if (!r) {
    return { ltp: null, bars: [], error: 'Public NIFTY: empty chart result' };
  }
  const meta = r.meta ?? {};
  const timestamps = r.timestamp;
  const q = r?.indicators?.quote?.[0] ?? {};
  const opens = q.open ?? [];
  const highs = q.high ?? [];
  const lows = q.low ?? [];
  const closes = q.close ?? [];

  /** @type {{ time: number, open: number, high: number, low: number, close: number }[]} */
  const bars = [];
  const len = Math.min(
    Array.isArray(timestamps) ? timestamps.length : 0,
    opens.length,
    highs.length,
    lows.length,
    closes.length,
  );
  for (let i = 0; i < len; i++) {
    const ts = timestamps[i];
    const open = finite(opens[i]);
    const high = finite(highs[i]);
    const low = finite(lows[i]);
    const close = finite(closes[i]);
    if (ts == null || open == null || high == null || low == null || close == null) continue;
    bars.push({
      time: ts > 1e12 ? ts : ts * 1000,
      open,
      high,
      low,
      close,
    });
  }

  let ltp = finite(meta.regularMarketPrice) ?? finite(meta.previousClose);
  if (bars.length) {
    const last = bars[bars.length - 1].close;
    ltp = last;
  } else if (Array.isArray(closes)) {
    for (let i = closes.length - 1; i >= 0; i--) {
      const c = finite(closes[i]);
      if (c != null) {
        ltp = c;
        break;
      }
    }
  }

  if (ltp == null) {
    return { ltp: null, bars: [], error: 'Public NIFTY: no price in response' };
  }

  const previousClose =
    finite(meta.chartPreviousClose) ?? finite(meta.previousClose) ?? null;

  return {
    ltp,
    bars,
    previousClose,
    error: '',
    note: 'NIFTY chart + headline from Yahoo ^NSEI (delayed). For broker LTP: npm run mstock:totp',
  };
}

async function fetchYahooChart() {
  const res = await fetch(CHART_URL, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (compatible; NiftyOptima/1.0; +local dev)',
      Accept: 'application/json',
    },
  });
  if (!res.ok) {
    return { ltp: null, bars: [], error: `Public NIFTY: HTTP ${res.status}` };
  }
  const json = await res.json();
  const r = json?.chart?.result?.[0];
  return parseYahooChartResult(r);
}

/**
 * @returns {Promise<{ ltp: number | null, error: string, note?: string }>}
 */
export async function fetchPublicNiftySpot() {
  try {
    const out = await fetchYahooChart();
    return { ltp: out.ltp, error: out.error, note: out.note };
  } catch (e) {
    return {
      ltp: null,
      error: e instanceof Error ? e.message : String(e),
    };
  }
}

/**
 * Spot and 1m bars from the same Yahoo request (keeps chart aligned with headline).
 * @returns {Promise<{ ltp: number | null, bars: { time: number, open: number, high: number, low: number, close: number }[], error: string, note?: string }>}
 */
export async function fetchPublicNiftyIntraday() {
  try {
    return await fetchYahooChart();
  } catch (e) {
    return {
      ltp: null,
      bars: [],
      error: e instanceof Error ? e.message : String(e),
    };
  }
}

function lastNBars(bars, n) {
  const sorted = [...bars].sort((a, b) => a.time - b.time);
  return sorted.slice(-n);
}

/**
 * Last N trading days from Yahoo ^NSEI (delayed).
 * @param {number} [tradingDays]
 */
export async function fetchPublicNiftyDaily(tradingDays = 5) {
  const days = Math.min(10, Math.max(1, Math.floor(tradingDays) || 5));
  try {
    const res = await fetch(DAILY_CHART_URL, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; NiftyOptima/1.0; +local dev)',
        Accept: 'application/json',
      },
    });
    if (!res.ok) {
      return { bars: [], error: `Public daily NIFTY: HTTP ${res.status}`, note: '' };
    }
    const json = await res.json();
    const r = json?.chart?.result?.[0];
    const out = parseYahooChartResult(r);
    const bars = lastNBars(out.bars, days);
    if (!bars.length) {
      return { bars: [], error: out.error || 'Public daily NIFTY: no bars', note: '' };
    }
    return {
      bars,
      error: '',
      note: 'Last 5 sessions from Yahoo ^NSEI (delayed). For broker data: log in via OTP.',
    };
  } catch (e) {
    return {
      bars: [],
      error: e instanceof Error ? e.message : String(e),
      note: '',
    };
  }
}

File: lib/niftyoptima/data/local/public_nifty_spot.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../../domain/entities/niftyoptima_models.dart';

/// Delayed NIFTY 50 from Yahoo ^NSEI when mStock Type B is unavailable.
class PublicNiftySpot {
  PublicNiftySpot({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;

  static const _chartUrl =
      'https://query1.finance.yahoo.com/v8/finance/chart/%5ENSEI?interval=1m&range=1d';

  Future<({
    double? ltp,
    List<MinuteBar> bars,
    double? previousClose,
    String error,
  })> fetchIntraday() async {
    try {
      final res = await _client
          .get(
            Uri.parse(_chartUrl),
            headers: const {
              'User-Agent': 'Mozilla/5.0 (compatible; NiftyOptima/1.0)',
              'Accept': 'application/json',
            },
          )
          .timeout(const Duration(seconds: 12));

      if (res.statusCode != 200) {
        return (
          ltp: null,
          bars: <MinuteBar>[],
          previousClose: null,
          error: 'Public NIFTY: HTTP ${res.statusCode}',
        );
      }

      final json = jsonDecode(res.body) as Map<String, dynamic>?;
      final result = (json?['chart'] as Map?)?['result'];
      if (result is! List || result.isEmpty) {
        return (
          ltp: null,
          bars: <MinuteBar>[],
          previousClose: null,
          error: 'Public NIFTY: empty chart',
        );
      }

      final r = result.first as Map<String, dynamic>;
      final meta = r['meta'] as Map<String, dynamic>? ?? {};
      final timestamps = r['timestamp'] as List? ?? [];
      final quote = (r['indicators'] as Map?)?['quote'];
      final q0 = quote is List && quote.isNotEmpty ? quote.first as Map? : null;
      final opens = (q0?['open'] as List?) ?? [];
      final highs = (q0?['high'] as List?) ?? [];
      final lows = (q0?['low'] as List?) ?? [];
      final closes = (q0?['close'] as List?) ?? [];

      final bars = <MinuteBar>[];
      final len = [timestamps.length, opens.length, highs.length, lows.length, closes.length]
          .reduce((a, b) => a < b ? a : b);

      for (var i = 0; i < len; i++) {
        final ts = timestamps[i];
        final open = _finite(opens[i]);
        final high = _finite(highs[i]);
        final low = _finite(lows[i]);
        final close = _finite(closes[i]);
        if (ts == null || open == null || high == null || low == null || close == null) {
          continue;
        }
        final time = ts is num ? (ts > 1e12 ? ts.toInt() : (ts * 1000).toInt()) : 0;
        bars.add(MinuteBar(time: time, open: open, high: high, low: low, close: close));
      }
      bars.sort((a, b) => a.time.compareTo(b.time));

      var ltp = _finite(meta['regularMarketPrice']) ?? _finite(meta['previousClose']);
      if (bars.isNotEmpty) {
        ltp = bars.last.close;
      } else {
        for (var i = closes.length - 1; i >= 0; i--) {
          final c = _finite(closes[i]);
          if (c != null) {
            ltp = c;
            break;
          }
        }
      }

      final previousClose =
          _finite(meta['chartPreviousClose']) ?? _finite(meta['previousClose']);

      if (ltp == null) {
        return (
          ltp: null,
          bars: bars,
          previousClose: previousClose,
          error: 'Public NIFTY: no price in response',
        );
      }
      return (ltp: ltp, bars: bars, previousClose: previousClose, error: '');
    } catch (e) {
      return (
        ltp: null,
        bars: <MinuteBar>[],
        previousClose: null,
        error: e.toString(),
      );
    }
  }

  double? _finite(dynamic v) {
    if (v is num && v.isFinite) return v.toDouble();
    return null;
  }
}

bool isMstockIpMismatch(String? text) {
  if (text == null || text.isEmpty) return false;
  return RegExp(
    r'IA403|ip address are not matching|Primary and Secondary IP',
    caseSensitive: false,
  ).hasMatch(text);
}

String mstockIpWhitelistHint() =>
    'Whitelist this device\'s public IP on trade.mstock.com → Trading APIs → '
    'Primary IP (check api.ipify.org on the phone). Mobile data uses a different IP than your PC.';
